File: aicq-crypto/src/keyExchange.ts
/**
 * X25519 ECDH shared-secret computation and session-key derivation.
 *
 * The HKDF implementation uses tweetnacl's SHA-512 hash (nacl.hash) to
 * approximate HKDF-SHA256.  This is *not* standard HKDF-SHA256, but it
 * provides a functional, self-contained key-derivation primitive that
 * only depends on tweetnacl (no Node.js crypto module required).
 */

import { nacl } from "./nacl.js";

/* ------------------------------------------------------------------ */
/*  ECDH shared secret                                                 */
/* ------------------------------------------------------------------ */

/**
 * Compute a 32-byte shared secret via X25519 Diffie-Hellman.
 */
export function computeSharedSecret(
  mySecretKey: Uint8Array,
  theirPublicKey: Uint8Array,
): Uint8Array {
  const shared = nacl.scalarMult(mySecretKey, theirPublicKey);
  if (!shared) {
    throw new Error("Key-exchange failed: scalarMult returned null");
  }
  return shared;
}

/* ------------------------------------------------------------------ */
/*  HKDF-like session-key derivation                                   */
/* ------------------------------------------------------------------ */

/**
 * Derive a 32-byte session key from a shared secret using an HKDF-like
 * construction built on top of nacl.hash (SHA-512).
 *
 * @param sharedSecret 32-byte X25519 shared secret.
 * @param context      Optional context / info string for domain separation.
 * @returns 32-byte session key.
 */
export function deriveSessionKey(
  sharedSecret: Uint8Array,
  context: string = "aicq-session",
): Uint8Array {
  // ---- Extract phase ----
  const salt = nacl.hash(
    new Uint8Array([...new TextEncoder().encode(context), ...sharedSecret]),
  );

  // ---- Expand phase ----
  // T(1) = HMAC-SHA512(salt, 0x01 || info)
  const infoBytes = new TextEncoder().encode(context);
  const expandInput = new Uint8Array([1, ...infoBytes]);
  const t1 = hmacSha512(salt, expandInput);

  // We only need 32 bytes, so truncate the 64-byte HMAC output.
  return t1.slice(0, 32);
}

/* ------------------------------------------------------------------ */
/*  Internal: HMAC-SHA-512                                             */
/* ------------------------------------------------------------------ */

/**
 * Minimal HMAC-SHA-512 built on top of nacl.hash (SHA-512).
 *
 * Since tweetnacl exposes a raw SHA-512 hash function (not HMAC), we
 * implement HMAC manually following RFC 2104.
 */
function hmacSha512(key: Uint8Array, message: Uint8Array): Uint8Array {
  const BLOCK_SIZE = 128; // SHA-512 block size in bytes

  // Ensure key is exactly one block long.
  let k: Uint8Array;
  if (key.length > BLOCK_SIZE) {
    k = nacl.hash(key); // 64 bytes
  } else {
    k = new Uint8Array(BLOCK_SIZE);
    k.set(key);
  }

  const oKeyPad = new Uint8Array(BLOCK_SIZE);
  const iKeyPad = new Uint8Array(BLOCK_SIZE);
  for (let i = 0; i < BLOCK_SIZE; i++) {
    oKeyPad[i] = k[i] ^ 0x5c;
    iKeyPad[i] = k[i] ^ 0x36;
  }

  const innerData = new Uint8Array(BLOCK_SIZE + message.length);
  innerData.set(iKeyPad);
  innerData.set(message, BLOCK_SIZE);
  const innerHash = nacl.hash(innerData);

  const outerData = new Uint8Array(BLOCK_SIZE + 64);
  outerData.set(oKeyPad);
  outerData.set(innerHash, BLOCK_SIZE);

  return nacl.hash(outerData);
}

File: shared/crypto/key_exchange.py
"""
X25519 key exchange and session key derivation.

Provides Diffie-Hellman shared secret computation via X25519 scalar
multiplication and an HKDF-like session key derivation using HMAC-SHA512,
matching the original @aicq/crypto TypeScript library.
"""

import hmac
import hashlib

import nacl.bindings
import nacl.hash
import nacl.encoding


def hmac_sha512(key: bytes, message: bytes) -> bytes:
    """Compute HMAC-SHA512.

    Args:
        key: The HMAC key.
        message: The message to authenticate.

    Returns:
        The 64-byte HMAC-SHA512 digest.
    """
    return hmac.new(key, message, hashlib.sha512).digest()


def compute_shared_secret(my_secret_key: bytes, their_public_key: bytes) -> bytes:
    """Compute an X25519 Diffie-Hellman shared secret.

    Performs scalar multiplication: shared = my_secret * their_public.

    Args:
        my_secret_key: Our 32-byte X25519 secret key.
        their_public_key: Their 32-byte X25519 public key.

    Returns:
        The 32-byte shared secret.

    Raises:
        ValueError: If key lengths are incorrect or scalar multiplication fails.
    """
    if len(my_secret_key) != 32:
        raise ValueError(
            f"Secret key must be 32 bytes, got {len(my_secret_key)}"
        )
    if len(their_public_key) != 32:
        raise ValueError(
            f"Public key must be 32 bytes, got {len(their_public_key)}"
        )

    try:
        return nacl.bindings.crypto_scalarmult(my_secret_key, their_public_key)
    except Exception as exc:
        raise ValueError(f"Scalar multiplication failed: {exc}") from exc


def derive_session_key(shared_secret: bytes, context: bytes = b"") -> bytes:
    """Derive a session key from a shared secret using HKDF-like construction.

    Uses HMAC-SHA512 in an extract-and-expand pattern:
    1. Extract: PRK = HMAC-SHA512(salt=context, input_key=shared_secret)
    2. Expand:  OKM = HMAC-SHA512(PRK, context || 0x01)

    Returns the first 32 bytes of the output keying material.

    Args:
        shared_secret: The 32-byte X25519 shared secret.
        context: Optional context/info bytes for domain separation.

    Returns:
        The 32-byte derived session key.
    """
    # Extract phase: PRK = HMAC-SHA512(context, shared_secret)
    prk = hmac_sha512(context, shared_secret)

    # Expand phase: OKM = HMAC-SHA512(PRK, context || 0x01)
    okm = hmac_sha512(prk, context + b"\x01")

    # Return first 32 bytes as the session key
    return okm[:32]
